kvserver: emit MVCC range tombstones over rangefeeds

erikgrinaker opened the pull request:
**storage: disable range keys in iterators when unsupported**

This patch falls back to `IterKeyTypePointsOnly` if the Pebble database
is not at `FormatRangeKeys` yet, to avoid a Pebble panic. These
databases can't contain range keys by definition, so this has no
practical effect.

Release note: None

**kvserver: emit MVCC range tombstones over rangefeeds**

This patch adds MVCC range tombstone support to range feeds. Whenever a
`DeleteRange` request with `UseExperimentalRangeTombstone` succeeds, a new
`MVCCDeleteRangeOp` logical op type is recorded and emitted across the
range feed as an `RangeFeedDeleteRange` event.

MVCC range tombstones are under active development, and highly
experimental. They will only be used (and emitted) after checking
`storage.CanUseExperimentalMVCCRangeTombstones`, which requires a
version gate and environment variable to be enabled.

Changefeeds will emit an error for these events. We do not expect to see
these in online spans with changefeeds, since they are initially only
planned for use with schema GC and import rollbacks.

The rangefeed client library has been extended with support for these
events, but no existing callers handle them for the same reason as
changefeeds. Initial scans do not emit regular tombstones, and thus not
range tombstones either, but catchup scans will emit them if
encountered.

This only has rudimentary integration testing, as MVCC range tombstones
are still experimental and lack e.g. persistence and replication -- more
comprehensive tests will be implemented later.

Resolves #70433.

Release note: None

erikgrinaker committed:
kvserver: emit MVCC range tombstones over rangefeeds

This patch adds MVCC range tombstone support to range feeds. Whenever an
MVCC range tombstone is written, e.g. a `DeleteRange` request with
`UseExperimentalRangeTombstone`, a new `MVCCDeleteRangeOp` logical op
type is recorded and emitted across the range feed as an
`RangeFeedDeleteRange` event.

MVCC range tombstones are under active development, and highly
experimental. They will only be used (and emitted) after checking
`storage.CanUseExperimentalMVCCRangeTombstones`, which requires a
version gate and environment variable to be enabled.

Changefeeds will emit an error for these events. We do not expect to see
these in online spans with changefeeds, since they are initially only
planned for use with schema GC and import rollbacks.

The rangefeed client library has been extended with support for these
events, but no existing callers handle them for the same reason as
changefeeds. Initial scans do not emit regular tombstones, and thus not
range tombstones either, but catchup scans will emit them if
encountered.

This only has rudimentary integration testing, as MVCC range tombstones
are still experimental and lack e.g. persistence and replication -- more
comprehensive tests will be implemented later.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76443)
<!-- Reviewable:end -->

erikgrinaker commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -380,6 +380,7 @@ func (s *eventStream) streamLoop(ctx context.Context, frontier *span.Frontier) e
 				}
 			default:
 				// TODO(yevgeniy): Handle SSTs.
+				// TODO(erikgrinaker): Handle DeleteRange events (MVCC range tombstones).
```
@miretskiy @dt Do we need to address these TODOs before stability in order to test streaming replication for 22.1? That is, add handling for `AddSSTable` and MVCC range tombstones in streaming replication.

miretskiy replied:
We decided that we will test what we can -- even if some of those ops are not ready yet.
So, these TODOs are fine for now.

erikgrinaker replied:
Ok. We have pending PRs/issues for the necessary APIs to ingest these on the target cluster, and plan to get these in for 22.1 in experimental form. We'll be maturing these APIs during stability and the 22.1 cycle. So if we get the streaming ingestion code in place before stability as well (i.e. simply make the appropriate API calls) then we should be able to do full end-to-end testing of streaming replication sometime during the 22.1 cycle.

miretskiy commented on pkg/kv/kvserver/rangefeed/catchup_scan.go:
```diff
@@ -94,6 +104,41 @@ func (i *CatchUpIterator) CatchUpScan(
+	// potentially discard point keys below them.
+	//
+	// Note that MVCC range tombstones are currently experimental, and will only
+	// be used when storage.CanUseExperimentalMVCCRangeTombstones() is enabled.
```
Trying to build a better mental model: the caller is responsible for building some sort of "frontier" so that it can filter out anything that falls in those ranges w/ lower timestamp?

erikgrinaker replied:
It's up to the caller how they want to handle these. The primary use-case we have now is streaming replication, which will simply ingest these range tombstones into the target cluster as-is, without doing anything fancy.

Some time in the future we may want to give e.g. SQL access to use these range tombstones when doing ranged deletions, at which point we'll have to figure out how to emit them across CDC -- possibly by simply emitting a new event type and leave it to the consumer to handle as appropriate. In principle, these are no different than point tombstones -- some consumers may want the past history below those points, while others may want to simply discard the old values. The same applies to range tombstones, with the caveat that we won't emit diffs for them.

miretskiy commented on pkg/kv/kvserver/rangefeed/catchup_scan.go:
```diff
@@ -94,6 +104,41 @@ func (i *CatchUpIterator) CatchUpScan(
+			}
+			i.rangeKeyIter.Next()
+		}
+	}
```
Do you think it would make sense to move this code into separate helper?  This function is pretty large already, imo.

erikgrinaker replied:
Yeah, good idea. Done.

miretskiy commented on pkg/roachpb/data.go:
```diff
@@ -2073,6 +2073,13 @@ func (u *LockUpdate) SetTxn(txn *Transaction) {
+func (s Span) Clone() Span {
+	s.Key = s.Key.Clone()
+	s.EndKey = s.EndKey.Clone()
+	return s
```
would it be better to  simply return `Span{Key: s.Key.Clone(), ...}` ?

erikgrinaker replied:
Yeah, definitely. Done.

miretskiy reviewed (commented):
few nits questions mostly.  I'm sure other reviewers will have more comments.

aliher1911 commented on pkg/kv/kvserver/rangefeed/catchup_scan.go:
```diff
@@ -242,3 +269,37 @@ func (i *CatchUpIterator) CatchUpScan(
 }
+
+// catchupScanRangeTombstones runs a catchup scan for range tombstones.
+func (i *CatchUpIterator) catchUpScanRangeTombstones(outputFn outputEventFn) error {
```
What is the reason for not filtering rangekeys on timestamp here? Point values are only sent when they are newer than catchup timestamp. Maybe that deserves a comment.

aliher1911 reviewed (approved):
LGTM

erikgrinaker committed:
storage: disable range keys in iterators when unsupported

This patch falls back to `IterKeyTypePointsOnly` if the Pebble database
is not at `FormatRangeKeys` yet, to avoid a Pebble panic. These
databases can't contain range keys by definition, so this has no
practical effect.

Release note: None

erikgrinaker commented:
Closing in favor of #82718.
